implementations/code_df/utils.py

```diff
@@ -66,3 +42,15 @@ def read_JSON_file(path):
 
             items.append(line)
     return items
+
+
+def rename_function(new_name):
```

Comment:
I see you're using this, but I don't understand why we need it.

Reply:
It is an attempt at generalizing the `compute_time_series` method. The `agg` function by default, sets the name of the aggregation method used as the name of the resulting column post aggregation. 
For example, 
```python
df = df['abc'].resample(..).agg(function_obj)
df.columns = df.columns.droplevel(0)
```
would make the df look like this:
```
              'function_obj`
index
..                ...
..                ...
..                ...
..                ...
```

By changing the name of the method, we do not need the user to pass in a separate argument to `compute_time_series` to name the aggregated column.

What do you think? Are there better ways to do it / is this not required?

Reply:
Another approach could be this:
Use a method to add a column specifically for the `time_series` or `compute` method, like this one in Code_Changes_Lines:
```python
    def _add_lines_modified_cols(self, df):

        if len(df) == 0:
            raise ValueError("DataFrame empty. "
                             "Please check instantiation parameters")

        additions = list()
        deletions = list()
        for _, commit in df.iterrows():
            added_lines = 0
            removed_lines = 0
            for file in commit['files']:
                if 'action' in file:
                    try:
                        added_lines += int(file['added'])
                        removed_lines += int(file['removed'])
                    except:
                        pass

            additions.append(added_lines)
            deletions.append(removed_lines)

        df['additions'] = additions
        df['deletions'] = deletions

        return df
```

And then we would not require renaming the method. Of course, `agg` would still use the method name (`_aggregate`) to reference a partcular column. To fix that, we can modify the `time_series` method in `metric.py` to this:
```python
        df = self.df

        df = df.set_index('created_date')
        # No particular column to aggregate on:
        if cols is None:
            cols = ['category']

        df = df[cols].resample(period).agg([self._aggregate])
        df.columns = df.columns.droplevel(1)
```

The result of `time_series` of CodeChangesLines would be something like this:
```python
code_changes_lines_obj.time_series(cols=['additions', 'deletions']
```
```
            additions            deletions
index 
..             ...                  ...
..             ...                  ....
..             ...                  ...
..             ...                  ...
```
where `additions` and `deletions` are just names of columns in `df` on which we applied the aggregation.
I think this method is better for more complex metrics, which would required aggregations on several columns. What do you think @jgbarah @aswanipranjal?

Reply:
The main problem is the naming of columns on performing aggregation.

Reply:
i sort of understand now. I would do it this way (assuming for each specific metric, we're going to use a single aggregation method, which i guess is the case):

* Have a method, say `_agg`, which would get a df as argument, and would return another df aggregated as you want. This method would be abstract in Metric / Commit, and will be defined for each of the specific metric.

* At the level we define the method `time_series`, (either in class Metric or in class Commit) you have the generic code, that calls `_agg()` when appropriate.

Does this sound ok to you?

Reply:
What I think you are saying is that `_agg` is not an aggregation method in the real sense. That is, it won't be passed to `Resampler.apply` or `Resampler.agg()`. Instead, it is just a regular method which takes a dataframe as input and returns whatever is required. Inside `_agg`, the resampling and aggregation happens.

The problem I think is that the aggregation requires passing a function object to the `agg` or `apply` method which is used after resampling:
```python3
df = df.resample('M').agg(my_agg_func)
```
And my_agg_func is passed a series: each column on which it is applied.
If we pass a dataframe to `_agg`, we would probably need another method to actually perform the aggregation (it will be passed to `Resampler.agg`).

The main motive behind the second method I suggested above is that for almost all of our metrics, we can add a few columns to `self.df` to allow for easier computation of metrics. Hence, we could have an overridden `_add_cols` method which adds columns to self.df. Also, whatever columns we added could be stored in a list.
Then, we could do this:
```python
        dict_ = {col: self._aggregate for col in self.added_cols}
        df = self.df
        df = df.set_index('created_date')
        df = df.resample(period).agg(dict_)
```
And self._aggregate would be: (for code_changes_lines)
```
    def _aggregate(self, series):
        total_modifications = series.sum()

        return total_modifications
```
What do you think?

Reply:
Sorry if this sounds like a very obvious question.

Reply:
I don't think we're understanding each other ;-) What I'm suggesting is rather simple. For the line that you have above:

```
df = df[cols].resample(period).agg([self._aggregate])
```

You would instead do something like:

```
df = self._agg(df)
```

And `_agg` would be like:

```
def _agg(self.df):

    return df.resample('M').count()
```

(assuming we're using months as the timeseries, and sum as the aggregating function. For a single metric, usually a single aggregation function is good enough. For example, .count() is the right one hen counting commits. Of course `_agg` for other metric could use a very different aggregation method.

Of course, you can also make this more general, defining a function which is the last one to be called on df.resample('M'), but maybe that's introducing an unneeded complexity. However, if you come with a clean solution, that could also be an option.

Is it more clear now?

Reply:
Definitely more clear! Will get back to you tomorrow morning.